docs/design/sql/20-sql-partition-pruning.md

```diff
@@ -0,0 +1,309 @@
+- `__key`
+
+_Clarification:_
+For IMap, partitioning columns may be attributes of IMap key. 
+If IMap uses `AttributePartitioningStrategy`, partitioning columns are functionally dependent on entire `__key`. 
+However for the sake of simplicity in the SQL optimizer we currently do not track functional dependencies
+and assume that both `__key` and `AttributePartitioningStrategy` define a partitioning key.
+Partition id calculation will take that into account and use a correct strategy for given IMap.
+
+*Definition: Partition-complete expression*
+
+Expression `E` is *partition-complete* when there exists partitioning key
+`PRK = {partColumn1, ..., partColumnM}`
+for which the expression can be transformed to form:
+`E = partKeyExp1 OR partKeyExp2 OR ... OR partKeyExpN`
+where `partKeyExpN` is in form 
+(there must be sub-expression for each of partitioning columns forming `PRK`):
+`partColumn1Expr AND partColumn1Expr AND ... AND partColumnMExpr AND residualFilter`
+where `partColumnMExpr` is in form
+`partColumnM <operator> <arguments>`
+where `operator` is one of the following operators:
+- `=`
+- `SEARCH` with `Sarg`
+- `BETWEEN` for integer types (TODO: if we want to support this in the first version, makes sense only for reasonably small range)
+- TODO: `IS NULL`?
+
+and operator `arguments` are:
+- literal
+- query parameter
+- complex expression using only the above `arguments` and any deterministic functions
+  (in particular references to columns are not allowed)
+
+and `residualFilter` is any expression, in particular it can be `TRUE`.  `residualFilter` may reference any columns,
+also the partitioning columns.
+
+Clarifications:
+- range partitioning (eg. `order_date between day1 and day1`) currently is not supported,
+  only hash/equality based partitioning is supported.
+- each `OR` branch in `E` might use different partitioning key. Currently, such case is not supported.
+- `partCol1 = partCol2 AND partCol2 = constantX` is not partition-complete, but can be transformed to such form
+  by inlining constant: `partCol1 = constantX AND partCol2 = constantX`. Such transformations are out-of-scope
+  for this TDD, but may be performed now or in future by the SQL optimizer independently.
+
+#### Goals
+
+- Improve performance of queries using IMap with `AttributePartitioningStrategy` which use attributes defined in it
+- Improve performance of queries using IMap `__key` which are not eligible to be converted to `IMapSelectPlan`
+  and have to create a Jet job
+
+#### Non-goals
+
+- Support partition pruning for other sources than IMap (can be considered in the future)
+- Complex expression transformations of filters to extract partitioning information
+  (for complex cases users will have to write their predicates in supported form)
+- Tracking of functional dependencies between attributes
+- Building a `__key` from constituent attributes, eg. if `__key` has 2 attributes, we could convert
```

Comment:
To my knowledge we don't, and it would be non-trivial to implement since we don't support OBJECT-ROW comparisons afaik. Without OBJECT-ROW comparisons we would need to construct the key object somehow, which is non-trivial too.